Do depth-first-search with new edge for cycle catching

lindsayad opened the pull request:
Per #20657 the dependency resolver is very slow when inserting new dependencies. In one of @jortensi's inputs I saw over 100 seconds spent in `RattlesnakeApp::addActionTasks`. With this change we do a depth-first search based on the new key we're adding in the model of https://www.geeksforgeeks.org/detect-cycle-direct-graph-using-colors/. This reduces the time spent in that method to .12 seconds.

lindsayad committed:
Do depth-first search to determine cylic dependencies

Refs #20657

lindsayad committed:
Remove isCylic method

loganharbour commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -226,25 +239,53 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
+  root_color_it->second = GRAY;
+
+  auto [first, last] = _depends.equal_range(root);
+  for (auto it = first; it != last; ++it)
+  {
+    const auto & adj = it->second;
```
```suggestion
  for (as_range(first, last))
  {
    const auto & adj = it.second;
```

cause why not? 😆

fdkong commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -226,25 +239,53 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
 
+template <typename T>
+bool
+DependencyResolver<T>::depthFirstSearch(const T & root)
```
Rename to "DetectCycle" Or "DetectCyclicDependency"?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1012470/) on 023bf73 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20778/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1012478/) on 023bf73 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>3587c9</th>
      <th colspan=4>#20778 023bf7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.56%</td>
      <td>82.57%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>74149</td>
      <td>74166</td>
      <td>+17</td>
      <td>20</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15661</td>
      <td>15661</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20778/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20778/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20778/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Document algorithm

Other changes:
- Rename `depthFirstSearch` to `hasCycle`
- Use `as_range` utility

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>
Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

permcody commented:
Oh like that this fancy pants fix!

lindsayad commented:
should be good for re-review

fdkong reviewed (dismissed):
Looks nice. Stack version might be even more interesting 

But we do not need to keep improving something that is already within 1 second

GiudGiud commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -226,25 +264,59 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
-
-    // Insert the breaking dependency here so it will reflect properly in the exception
-    depends_copy.insert(k);
+    _depends.erase(insert_it);
```
is this needed? We're throwing anyway

lindsayad replied:
Just because you throw doesn't mean the simulation is over. That's the whole reason to throw vs abort

GiudGiud commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -226,25 +264,59 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
+  auto insert_it = _depends.insert(k);
 
-  if (dependsOn(value, key))
+  for (auto & pr : _colors)
+    pr.second = WHITE;
```
I think you'd only need to reset the connected component. But this is plenty of speedup already :)

lindsayad replied:
> component

Do you mean components? Adding a new edge could complete a cycle involving an arbitrarily large number of edges. You would have to reset the color of all the nodes in that cycle in order for this algorithm to detect the cycle. Are you suggesting adding a new method that descends the tree, coloring all nodes it finds WHITE, and then run the `hasCycle` method?

GiudGiud replied:
yeah I mean components. Like starting from this node to every reachable node.

This could be slower than simply traversing the entirety of colors. So I dont know!

lindsayad replied:
I think for resetting colors, what you would have to do is descend from the `value` node pre-new-edge-insertion since the insertion can in theory create a cycle and then you would be descending forever. We could try that at some point and check to see whether it's faster than simply iterating over the container which will be guaranteed to be a linear cost directly proportional to the total size of the `_colors` container. Selectively marking should be proportional to the size of the descent chain which will obviously be strictly less or equal to the size of the `_colors` container ... so it could be faster.

GiudGiud commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -132,7 +135,37 @@ class DependencyResolver
+   * GRAY: Node is being processed (DFS for this node has started, but not finished which means that
+   * all descendants (in DFS tree) of this node are not processed yet (or this node is in the
+   * function call stack)
+   * BLACK : Node and all its descendants are processed. While doing DFS, if an edge is
+   * encountered from current node to a GRAY node, then this edge is back edge and hence there
+   * is a cycle.
```
```suggestion
   * BLACK : Node and all its descendants are processed.
   * If an edge is encountered from current node to a GRAY node, then this edge is 
   * back edge and hence there is a cycle.
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1013929/) on 54a127e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20778/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20778/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 742250a58142ca80b131fe15c4a14c50c8bb39a6`

lindsayad commented:
ugh I should have just merged this

GiudGiud commented on framework/include/utils/DependencyResolver.h:
```diff
@@ -226,25 +262,59 @@ DependencyResolver<T>::insertDependency(const T & key, const T & value)
-    depends_copy.insert(k);
+    _depends.erase(insert_it);
 
     throw CyclicDependencyException<T>(
```
@lindsayad @lingzou 
SAM is finding a new cyclic dependency, how do we return the name of the key here? It would help them
I imagine with the template _console << key or similar wont work since it's not defined. can we do some sort of template enable_if inside a routine? or should we duplicate the function ?
